src/persist-client/src/internal/state.rs

```diff
@@ -1401,4 +1544,12 @@ mod tests {
             })
         );
     }
+
```

Comment:
wouldn't mind a test or two that explicitly demonstrate how writes can only advance `writer_state.most_recent_write_upper` + that idempotency tokens are recorded correctly

Reply:
didn't get to this one before I had to stop for the day. will do it tomorrow

Reply:
didn't actually do anything that prints out the writer states at various points of the test because I wasn't sure it was worth Yet Another Datadriven Directive. happy to add it if you decide you still want something like that after looking at what I did